### ISSUE TITLE
[FEA] Try a sort-free coo2csr to reduce coo2csr memory requirement on SNSG

### ISSUE THREAD

**Issue by afender** [rapidsai/cugraph#153]
**Is your feature request related to a problem? Please describe.**
We identified that cugraph’s peak memory requirement is in coo2csr. The current implementation sorts the edge list and then builds the CSR matrix

**Describe the solution you'd like**
The sort needs a temporary buffer equivalent ~2 edge lists. Andrei has a sort-free GPU implementation that uses a simple form of hashing. 

**Describe alternatives you've considered**
- UVM in RMM
- Manual memory management + sort that can stream data (much more complex to develop)
- Use Multiple GPUs (long term solution - need something sooner)

**Task List**

- [ ] get a memory profile 
- [ ] get performance results on large datasets 
- [ ] compare (time and memory) to what we have currently in cuGraph

(if better than what we have in cugraph)
- [ ] integrate in cugraph
- [ ] run all tests and verify that there is no unexpected regression

**Comment by ChuckHastings** [comment]
Current status of my coo2csr exploration.  For my arithmetic below, n = size (the number of vertices - technically the maximum vertex id), m = nnz (the number of edges).

The original code:
 * makes a copy of source and destination arrays (since we can't corrupt the original)
 * sorts the copies to order the COO, then computes the starting point for the CSR.

This uses ~ m + 3n extra memory (one of the n's is an estimate guessed from what the CUB call seems to require) that is freed at the end of the call.


I implemented a coo2csr that does not require sorting the COO.  This prevents the need to copy the source and destination array.  My algorithm constructs the CSR in a non-deterministic way directly from the original arrays.  It then does a segmented sort of the resulting array (segmented sort should be faster than a full sort - theoretically).

Construction of the unsorted CSR is markedly faster than the original algorithm.  If we don't need the edge lists to be sorted, the technique works very well.  When adding the segmented sort - at least for now - the total cost is comparable (about 10-20% slower).

The one exception seems to be the rmat graph that I generated for testing.  In this case both the creation of the unsorted CSR and the sorting are markedly slower.

There is a performance enhancement for constructing the CSR that I haven't put in yet.  The segmented sort still requires tuning.

My algorithm only requires an additional (1 + n) memory, everything else is done in place - in the output arrays so that the original input is preserved.

It seems like my algorithm would translate to SNMG, but I haven't specifically laid out how.
